### DISCUSSION TITLE
react-hook-form

### DISCUSSION THREAD

**Discussion by jin-Pro** [AjouCapstone-Sw/client#4] [Ideas]
## react-hook-form 이란?

**Comment by Noelsky-code** [comment]
[공식 사이트](https://react-hook-form.com/)

react-hook-form은 로그인, 회원가입, 글 작성 과 같이 __여러가지 데이터를 입력__받아야 하는, 리액트에서의 form태그 사용을 __효율적__으로 해주는 라이브러리입니다.

react-hook-form은 아래와 같은 특징을 가지고 있으며 사용하게 된다면 생산성, 관리에 큰 도움이 될 거라 기대됩니다 

### input 상태 관리 
리액트를 사용하며 사용자에게 여러 입력 값을 입력받을 때 각 입력값에 대한 state를 생성해 주고 관리해줘야하는 번거로움은 당연히 느꼇을 것이라 생각됩니다.
reducer를 사용해도 각 입력 값에 대해 액션들을 정의해 줘야한다는 번거로움은 여전히 남습니다. 

react-hook-form을 사용하면 이런 불편함을 해소시킬 수 있으며 기존 코드량을 크게 줄여 생산성을 크게 증가시킬 수 있습니다. 
react-hook-form에서는 아래 코드와 같이 form에서 사용되는 데이터 타입을 미리 지정하고, input 태그에 등록해주면 됩니다. 

```javascript 
interface IAuthForm {
  nickname: string;
  email: string;
  pw: string;
}

... 

const {
    register,
    handleSubmit,
  } = useForm<IAuthForm>();

const onSubmit = (data)=> ...
...
return (
    <form onSubmit={handle(onSubmit)}>
      <fieldset>
        <label htmlFor="nickname">닉네임</label>
        <input {...register("nickname")} />
        <label htmlFor="email">이메일</label>
        <input {...register("email")} />
        <label>비밀번호</label>
        <input {...register("pw")} />
      </fieldset>
    </form>
  )

```

### input validation
react-hook-form의 가장 큰 장점 중 하나는 간단한 input validation입니다.
기본적으로 react에서 입력 값에 대한 validation은 상태를 통해 이루어집니다.
input 값에 대한 validation 체크는 
1. 입력 값 변할 때 마다 validation 체크 함수 실행
2. validation이 유효하지 않을 때 error state 변경 
3. do something 

하지만 react-hook-form에서는 위 과정을 props 전달을 통해 간단하게 해결할 수 있습니다 

```
  <input {...register("email", {
        required:"필수 응답 항목입니다.", 
        pattern:{
          value:/^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/i, 
          message:"이메일 형식이 아닙니다."}}
        ,minLength:{
               value:8,
               message: "비밀번호는 8자 이상이어야합니다"
         }
/>
```
만약 위 코드를 react-hook-forom을 사용하지 않고 구현했다면, 
이메일 형식에 대한 error state, 길이에 대한 error state, 
패턴 체크, 길이 체크에 대한 함수, error msg를 띄우는 함수가 추가로 필요합니다.
react-hook-form을 사용하면 위 코드와 같이 간략하게 작성이 가능하며 더욱 직관적이며 생산적입니다

### 성능적인 강점 
기존의 리액트에서 input 상태를 여러개 관리 하는 방법을 다시 생각해 보면,
input 하나의 state가 변경되면 모든 자식 컴포넌트가 리렌더링됩니다.

하지만 react-hook-form는 비제어 컴포넌트 방식으로 구현되어 있기 때문에 자식 컴포넌트의 리렌더링이 발생하지 않습니다. 그리고 페이지에 컴포넌트가 마운트되는 속도도 매우 빠릅니다. 

또한, 다른 dependency를 가지지 않는 라이브러리로 라이브러리 크기가 매우 작습니다.

**Reply by jin-Pro** [reply]
훌륭합니다.

wiki에 그대로 이관해도 됄 내용인 것 같아요.

매번 값에 대한 validation 코드를 작성하느라 코드가 지저분해져 자주 분리하는 로직을 작성했는데, 매우 간단하네요.

좋은 자료 감사합니다. 더 찾아보지 않아도 됄 것 같네요.
